altair/utils/data.py

```diff
@@ -2,63 +2,84 @@
 import os
 import random
 import hashlib
+import sys
 import warnings
+from typing import Union, MutableMapping, Optional, Dict, Sequence, TYPE_CHECKING, List
+from types import ModuleType
 
 import pandas as pd
 from toolz import curried
-from typing import Callable
+from typing import TypeVar
 
 from .core import sanitize_dataframe, sanitize_arrow_table
 from .core import sanitize_geo_interface
 from .deprecation import AltairDeprecationWarning
 from .plugin_registry import PluginRegistry
 
 
+if sys.version_info >= (3, 8):
+    from typing import Protocol, TypedDict, Literal
+else:
+    from typing_extensions import Protocol, TypedDict, Literal
+
+
+if TYPE_CHECKING:
+    import pyarrow.lib
+
+
+class _SupportsGeoInterface(Protocol):
+    __geo_interface__: MutableMapping
+
+
+class _SupportsDataframe(Protocol):
+    def __dataframe__(self, *args, **kwargs):
+        ...
```

Comment:
We can replace this with `_DataFrameLike` now:

https://github.com/altair-viz/altair/blob/f3938bff6b32446623ad56455224c9b341a0f066/altair/utils/core.py#L40-L42